server/lib/tuist/qa/agent.ex

```diff
@@ -3,33 +3,212 @@ defmodule Tuist.QA.Agent do
   Tuist QA agent module.
   """
 
+  alias LangChain.Chains.LLMChain
+  alias LangChain.ChatModels.ChatAnthropic
+  alias LangChain.Message
+  alias LangChain.Message.ContentPart
+  alias LangChain.Message.ToolResult
+  alias LangChain.TokenUsage
+  alias Tuist.QA.Tools
   alias Tuist.Simulators
   alias Tuist.Zip
 
-  def test(%{preview_url: preview_url, bundle_identifier: bundle_identifier}) do
-    run_preview(preview_url, bundle_identifier)
+  require Logger
+
+  def test(%{preview_url: preview_url, bundle_identifier: bundle_identifier, prompt: prompt}) do
+    with {:ok, simulator_device} <- simulator_device(),
+         :ok <- run_preview(preview_url, bundle_identifier, simulator_device) do
+      handler = %{
+        on_message_processed: fn _chain, %Message{content: content} ->
+          for %ContentPart{type: :text, content: content} <- content || [] do
+            Logger.debug(content)
+          end
+        end,
+        on_llm_token_usage: fn _chain, %TokenUsage{input: input, output: output} ->
+          Logger.debug("LLM token usage: #{input} input tokens, #{output} output tokens")
+        end
+      }
+
+      prompt = """
+      You are a QA agent. Test the following: #{prompt}.
+
+      First, understand what you need to test. Then, set up a plan for test the feature and execute on it without asking for additional instructions. Make sure to test **multiple** scenarios. Try to interact with most of the active elements in the feature that should be tested. Take screenshots to analyze whether there are any visual inconsistencies and store the screenshots at /tmp/screenshots.
+
+      For each step, return the test result output as a JSON in the form of {title: summary of what you tested, description: deeper explanation of what you tested and why, issues: [issues that you found], screenshots: [paths of screenshots that you took]}. Wrap this JSON in ```json``` so we can easily parse it. When using tools, use the following udid: #{simulator_device.udid}.
```

Comment:
regarding the individual steps, I might use a dedicated tool. I will work on this in a follow-up that will send the individual steps and screenshots to the server, so this piece is very much subject to change.

Reply:
I'm gonna do that in a follow-up 👍